Improve XKRXExchangeCalendar

elbakramer opened the pull request:
Core changes in `ExchangeCalendar `

* [Can give `None` to break times](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L126-L139) in order to remove break time since the given date
* [Can configure special weekmasks](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L152-L156) that are applied to specific date range only
* [Can configure special offsets](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L168-L217) that are applied to specific date range only


`XKRXExchangeCalendar` specific changes

* `XKRXExchangeCalendar` is no longer `PrecomputedExchangeCalendar` but `ExchangeCalendar`
* Complies with [full historical schedule changes](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L119-L145) with possible historical break times
* Have [special historical weekmask](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L152-L156) which includes Saturday as business day until 1998-12-07
* Have [special offsets](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L192-L217) that delays all schedules by an hour (or by 30 minutes before 1998-11-18) on [CSAT days](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/xkrx_holidays.py#L1194-L1231)
* Have [special offsets](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L171-L190) that delays opening time by an hour on the first business day of year
* Holidays are now maintained in a separate file [`xkrx_holidays.py`](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/xkrx_holidays.py)
* Have [regular holidays](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L160-L162) (including handling korean lunar holidays) which can extrapolate until 2050-12-31
* Have [adhoc holidays](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/exchange_calendar_xkrx.py#L164-L166) updated based on [the official krx website reference](http://global.krx.co.kr/contents/GLB/05/0501/0501110000/GLB0501110000.jsp)
* Adhoc holidays can be updated automatically using a [utility script](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/etc/update_xkrx_holidays.py#L121)

Minor changes

* Fix [`FutureWarning` for indexing timezone-naive dt to timezone-aware index](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/ecal.py#L62) in [`ecal.py`](https://github.com/elbakramer/exchange_calendars/blob/fc716785bca85b8a33642cb8fc91e0e928b1bcf8/exchange_calendars/ecal.py#L44)

gerrymanoim commented:
Apologies - I know I owe you actual feedback here but have been quite busy lately. Trying to check in keep the CI running though.

elbakramer commented:
It's OK.

I just figured out that I also can activate the Github Actions in my fork and run CI tests on push.
So I'll let you know when all failing checks are fixed.

It seems that I'm still missing many checks btw...

elbakramer commented:
Finally all checks have passed in my fork: https://github.com/elbakramer/exchange_calendars/runs/2479460122

gerrymanoim commented on exchange_calendars/pandas_extensions/holiday.py:
```diff
@@ -0,0 +1,151 @@
+            end_date,
+            days_of_week,
+        )
+        self.name = name
+        self.year = year
+        self.month = month
+        self.day = day
+        self.offset = offset
+        self.start_date = (
+            Timestamp(start_date) if start_date is not None else start_date
+        )
+        self.end_date = Timestamp(end_date) if end_date is not None else end_date
+        self.observance = observance
+        assert days_of_week is None or type(days_of_week) == tuple
+        self.days_of_week = days_of_week
```
This seems the same as https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/holiday.py#L214-L225. 

You should be able to get away with just 

```
self.offset = offset
self.observance = observance
```
after your super call right?

elbakramer replied:
You're right. I just copy/pasted the codes at the first time, and then put that super call later.
I think below change would suffice, as you pointed out.

```suggestion
        self.offset = offset
        self.observance = observance
```

gerrymanoim commented on exchange_calendars/xkrx_holidays.py:
```diff
@@ -0,0 +1,1231 @@
+
+
+# Automatically generated holidays using /etc/update_xkrx_holidays.py script.
+# Note that there are some missing holidays compared to the original holidays.
```
Any idea why they're missing?

elbakramer replied:
Not much actually.

The following page offers holidays ranging from current year to year 2009:
http://global.krx.co.kr/contents/GLB/05/0501/0501110000/GLB0501110000.jsp
And it seems that those holidays are obviously valid with no missing holidays.

While it might seem that we can only check the holidays until year 2009 in the page,
that page's underlying API even gives responses with years older than 2009, the limitation in UI.
So I just blindly dumped the responses from that API as much as I can get and combined them with the existing ones.

I expected that it had to be the single source of truth but just turned out that it was not, with some missing holidays.
Since this is not an desirable usage in their perspective, I think we cannot expect any of integrity in data beyond year 2009.

gerrymanoim commented on exchange_calendars/exchange_calendar_xkrx.py:
```diff
@@ -569,18 +64,195 @@ class XKRXExchangeCalendar(PrecomputedExchangeCalendar):
+        ]
+
+
+class PrecomputedXKRXExchangeCalendar(PrecomputedExchangeCalendar):
```
Any reason to keep this one around?

elbakramer replied:
For those who may want the original behavior for any reason.

gerrymanoim commented on exchange_calendars/exchange_calendar_xkrx.py:
```diff
@@ -569,18 +64,195 @@ class XKRXExchangeCalendar(PrecomputedExchangeCalendar):
+
+    @property
+    def adhoc_holidays(self):
+        return precomputed_krx_holidays.tolist()
```
What's the overlap between the precomputed holiday and the holidays generated by the new rules? Or are they disjoint?

elbakramer replied:
Since the original precomputed holidays covered both regular and non-regular (adhoc) holidays when the calendar was `PrecomputedExchangeCalendar`, each of newly suggested regular holidays (generated by the new rules) should generally fall into one of the original precomputed holidays (except the future holidays that are not precomputed yet).

So they are not disjoint. Possibly the generated holidays should be a subset of precomputed holidays.

Things to note however:

1. precomputed holidays pre-emptively exclude weekends (saturdays and sundays) while the generated holidays do not.
2. precomputed holidays may include adhoc holidays such as election day while the generated holidays cannot cover them.
3. still precomputed holidays may omit some valid holidays and generated holidays might cover those missing holidays.
4. and still there might be some valid adhoc holidays that both precomputed and generated holidays miss out.

After listening to your question, I quickly added a test to check whether my argument is right (with additional fix to labor day). Currently that test fails but I think those cases all fall into the case 3 explained above.

<details><summary>Generated holidays that don't fall into the precomputed holidays</summary>
<p>

```
>       self.assertTrue(all(isin), "missing holidays = \n%s" % missing)
E       AssertionError: False is not true : missing holidays =
E       1975-09-19        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1976-01-30    Seollal (New Year's Day by the lunar) (-1 day)
E       1976-09-07        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1976-09-09        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1977-02-17    Seollal (New Year's Day by the lunar) (-1 day)
E       1977-02-18             Seollal (New Year's Day by the lunar)
E       1977-09-26        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1977-09-28        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1978-02-06    Seollal (New Year's Day by the lunar) (-1 day)
E       1978-02-07             Seollal (New Year's Day by the lunar)
E       1978-02-08    Seollal (New Year's Day by the lunar) (+1 day)
E       1978-09-18        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1979-01-29    Seollal (New Year's Day by the lunar) (+1 day)
E       1979-10-04        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1980-02-15    Seollal (New Year's Day by the lunar) (-1 day)
E       1980-09-22        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1980-09-24        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1981-02-04    Seollal (New Year's Day by the lunar) (-1 day)
E       1981-02-05             Seollal (New Year's Day by the lunar)
E       1981-02-06    Seollal (New Year's Day by the lunar) (+1 day)
E       1981-09-11        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1982-01-25             Seollal (New Year's Day by the lunar)
E       1982-01-26    Seollal (New Year's Day by the lunar) (+1 day)
E       1982-09-30        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1983-02-14    Seollal (New Year's Day by the lunar) (+1 day)
E       1983-09-20        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1983-09-22        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1984-02-01    Seollal (New Year's Day by the lunar) (-1 day)
E       1984-02-02             Seollal (New Year's Day by the lunar)
E       1984-02-03    Seollal (New Year's Day by the lunar) (+1 day)
E       1984-09-11        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1985-02-19    Seollal (New Year's Day by the lunar) (-1 day)
E       1985-02-21    Seollal (New Year's Day by the lunar) (+1 day)
E       1985-09-30        Chuseok (Korean Thanksgiving Day) (+1 day)
E       1986-02-10    Seollal (New Year's Day by the lunar) (+1 day)
E       1986-09-17        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1987-01-28    Seollal (New Year's Day by the lunar) (-1 day)
E       1987-01-30    Seollal (New Year's Day by the lunar) (+1 day)
E       1987-10-06        Chuseok (Korean Thanksgiving Day) (-1 day)
E       1988-02-17    Seollal (New Year's Day by the lunar) (-1 day)
E       1988-02-19    Seollal (New Year's Day by the lunar) (+1 day)
E       1996-02-19             Seollal (New Year's Day by the lunar)
E       1996-02-20    Seollal (New Year's Day by the lunar) (+1 day)
E       1997-05-01                                         Labor Day
E       1997-05-14                                 Buddha's Birthday
```

</p>
</details>

I'm thinking of putting those holidays to precomputed list or just limiting the testing date range not to include those dates, when I'm confident with my argument.

Some external references for that:

1. [1997-05-01 is holiday because it's Labor day](https://www.hankyung.com/finance/article/1997042802391)
2. [All the dates above are missing in historical daily quotes of a symbol](https://finance.daum.net/quotes/A000020#current/quote)
3. [1997-12-18 is holiday for election but both precomputed and generated are missing out this holiday](https://www.hankyung.com/finance/article/1997121600791)

gerrymanoim reviewed (commented):
I took a quick look at these changes today. Thanks for all the work! Very impressive. I left a few quick notes, but in general I think this looks reasonably good. 

I'm a bit wary of subclassing the pandas Holiday/Businessday/Calendar objects (both from a maintainability and using this library in other libraries perspective) but I think I'm okay using it on just the XKRX calendar and what you've done seems reasonable. I'd like to take a deeper look at those a bit later though.

elbakramer commented:
Some fixes:

1. [Fixed `KoreanHoliday`'s alternative holiday behavior](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/pandas_extensions/korean_holiday.py#L21-L25) which was not working at all in fact
2. [Fixed `CompositeCustomBusinessDay`'s `is_on_offset()` method](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/pandas_extensions/offsets.py#L173-L178) which disregarded overrided bdays
3. Added some missing precomputed/adhoc holidays
    * [1996-02-19: Seollal](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L174)
    * [1996-02-20: Seollal (+1)](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L175)
    * [1997-05-01: Labor Day](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L191)
    * [1997-05-14: Budda's Birthday](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L193)
    * [1997-12-18: Election Day (adhoc)](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L201)
4. Added some missing historical regular holidays
    * [additional holidays after New Years Day](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1100-L1113)
    * [Arbor Day](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1140-L1146)
    * [Constitution Day](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1158-L1164)
    * [Armed Forces Day](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1189-L1195)
    * [(Old) Hangul Proclamation Day (until once excluded from holidays temporarily)](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1199-L1205)
5. Added missing start/end date limits for some of the existing regular holidays
    * [-1/+1 day holidays of Seollal](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1114-L1136) and [Chuseok](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1166-L1188)
6. [Fixed end of year holiday's missing observance behavior](https://github.com/elbakramer/exchange_calendars/blob/108abb949693a133d28d0a164914daaac62502d9/exchange_calendars/xkrx_holidays.py#L1218)

gerrymanoim commented:
Apologies - I'll take a look again and get this merged this weekend.

elbakramer commented:
It's OK. Take your time.

gerrymanoim commented:
@elbakramer - Thanks for all the work on this. Merged now.
